pallets/gear/src/manager/mod.rs

```diff
@@ -388,4 +365,80 @@ where
             log::trace!("Signal wasn't sent due to inappropriate supply");
         }
     }
+
+    /// Removes memory pages of the program and transfers program balance to the `value_destination`.
+    fn clean_inactive_program(program_id: ProgramId, value_destination: ProgramId) {
+        ProgramStorageOf::<T>::remove_program_pages(program_id);
+
+        let program_account = &<T::AccountId as Origin>::from_origin(program_id.into_origin());
+        let balance = CurrencyOf::<T>::free_balance(program_account);
+        if !balance.is_zero() {
+            let destination = Pallet::<T>::inheritor_for(value_destination);
+            let destination = <T::AccountId as Origin>::from_origin(destination.into_origin());
+
+            CurrencyOf::<T>::transfer(
+                program_account,
+                &destination,
+                balance,
+                ExistenceRequirement::AllowDeath,
+            )
+            .unwrap_or_else(|e| unreachable!("Failed to transfer value: {e:?}"));
+        }
+    }
+
+    /// Removes all messages to `program_id` from the waitlist.
+    fn clean_waitlist(program_id: ProgramId) {
+        let reason = MessageWokenSystemReason::ProgramGotInitialized.into_reason();
+
+        WaitlistOf::<T>::drain_key(program_id).for_each(|entry| {
+            let message = Pallet::<T>::wake_dispatch_requirements(entry, reason.clone());
+
+            QueueOf::<T>::queue(message)
+                .unwrap_or_else(|e| unreachable!("Message queue corrupted! {e:?}"));
+        });
+
+        ProgramStorageOf::<T>::waiting_init_remove(program_id);
+    }
+
+    fn process_failed_init(program_id: ProgramId, origin: ProgramId, executed: bool) {
+        // Some messages addressed to the program could be processed
+        // in the queue before init message. For example, that could
+        // happen when init message had more gas limit then rest block
+        // gas allowance, but a dispatch message to the program was
+        // dequeued. The other case is async init.
+        Self::clean_waitlist(program_id);
+
+        ProgramStorageOf::<T>::update_program_if_active(program_id, |p, bn| {
```

Comment:
do we still have invariant that journal note exit goes after message_dispatched? what if I wait init message but call exit in handle_reply?

Reply:
Opened issue. Will assign for you @gshep